### PR TITLE
Added configuration, resources, and APIs for AAP.

### PR DESCRIPTION
**Pull request by russjones**
**Description**

Updated file configuration to add support for `app_service` and propagate configuration to `service` package.

Added `services.App` resource (represents an application) and CRUD API to GRPC server. Also updated cache to support fast access to `GetApp` and `GetApps`.

Updated `services.Role` to add `app_labels` to the role conditions. Also added `CheckAccessToApp` for RBAC checks.

Updated permissions to allow proxy to read `services.KindApp` types of resources.

**Related Issues**

https://github.com/gravitational/teleport/issues/3951

### CODE REVIEW

File: lib/auth/auth_with_roles.go [hunk 1]
```diff
@@ -519,6 +523,36 @@ NextNode:
...
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Loop over all apps and check if the caller has access.
+	filteredApps := make([]services.App, 0, len(apps))
+NextApp:
+	for _, app := range apps {
+		err := roleset.CheckAccessToApp(app)
+		if err == nil {
```
Comment (awly) [inline lib/auth/auth_with_roles.go:547]:
```
if err != nil {
  if trace.IsAccessDenied(err) {
    continue
  }
  return nil, trace.Wrap(err)
}
```

File: lib/services/suite/suite.go [hunk 2]
```diff
@@ -356,6 +356,47 @@ func (s *ServicesTestSuite) ServerCRUD(c *check.C) {
...
+		Spec: services.AppSpecV3{
+			Protocol:     teleport.WebAppProtocol,
+			InternalAddr: internalAddr,
+			PublicAddr:   publicAddr,
+		},
+	}
+}
+
+// AppCRUD tests CRUD functionality for services.App.
+func (s *ServicesTestSuite) AppCRUD(c *check.C) {
```
Comment (awly) [inline lib/services/suite/suite.go:377]: Also check `GetApp`

File: lib/services/types.proto [hunk 3]
```diff
@@ -975,3 +984,41 @@ message TunnelConnectionSpecV2 {
...
+    option (gogoproto.goproto_stringer) = false;
+    option (gogoproto.stringer) = false;
+
+    // Kind is a resource kind
+    string Kind = 1 [ (gogoproto.jsontag) = "kind" ];
+    // SubKind is an optional resource sub kind, used in some resources
+    string SubKind = 2 [ (gogoproto.jsontag) = "sub_kind,omitempty" ];
+    // Version is version
+    string Version = 3 [ (gogoproto.jsontag) = "version" ];
+    // Metadata is User metadata
```
Comment (awly) [inline lib/services/types.proto:999]: I really wish these were embedded and reusable by all types

**Review by awly — COMMENTED** [review]
That's a lot of plumbing required for adding a new CRUD type.
If we do this more often, we should think about codegen or something.

Could you also add a unit test for config parsing/validation?

File: lib/auth/permissions.go [hunk 4]
```diff
@@ -276,6 +276,7 @@ func GetCheckerForBuiltinRole(clusterName string, clusterConfig services.Cluster
 							services.NewRule(services.KindGithubRequest, services.RW()),
 							services.NewRule(services.KindNamespace, services.RO()),
 							services.NewRule(services.KindNode, services.RO()),
+							services.NewRule(services.KindApp, services.RO()),
```
[the remark this answers is not in the document]
Comment (russjones) [reply lib/auth/permissions.go:300]: Proxies can not, only App nodes can.

File: lib/auth/proto/auth.proto [hunk 5]
```diff
@@ -248,4 +297,15 @@ service AuthService {
...
+    // GetApp returns an applications.
+    rpc GetApp(GetAppRequest) returns (GetAppResponse);
+    // GetApps returns all registered applications.
+    rpc GetApps(GetAppsRequest) returns (GetAppsResponse);
+    // UpsertApp registers an application with a TTL.
+    rpc UpsertApp(UpsertAppRequest) returns (services.KeepAlive);
+    // DeleteApp un-registers an application.
+    rpc DeleteApp(DeleteAppRequest) returns (google.protobuf.Empty);
+    // DeleteAllApp un-registers an application.
+    rpc DeleteAllApps(DeleteAllAppsRequest) returns (google.protobuf.Empty);
```
[the remark this answers is not in the document]
Comment (russjones) [reply lib/auth/proto/auth.proto:310]: Never surfaced to user.

File: lib/cache/collections.go [hunk 6]
```diff
@@ -1051,3 +1056,66 @@ func (c *role) processEvent(ctx context.Context, event services.Event) error {
...
+
+func (a *app) processEvent(ctx context.Context, event services.Event) error {
+	switch event.Type {
+	case backend.OpDelete:
+		err := a.presenceCache.DeleteApp(ctx, event.Resource.GetMetadata().Namespace, event.Resource.GetName())
+		if err != nil {
+			// Resource could be missing in the cache expired or not created, if the
+			// first consumed event is delete.
+			if !trace.IsNotFound(err) {
+				a.Warningf("Failed to delete resource %v.", err)
```
[the remark this answers is not in the document]
Comment (russjones) [reply lib/cache/collections.go:1100]: Copied from code @fspmarshall wrote, can you we remove this warning?

File: lib/config/fileconf.go [hunk 7]
```diff
@@ -778,6 +786,42 @@ func (b *BPF) Parse() *bpf.Config {
...
+type App struct {
+	// Name of the application.
+	Name string `yaml:"name"`
+
+	// Protocol used to proxy this application. At the moment only HTTPS is
+	// supported.
+	Protocol string `yaml:"protocol"`
+
+	// URI is the internal address of the application.
+	URI string `yaml:"uri"`
```
[the remark this answers is not in the document]
Comment (russjones) [reply lib/config/fileconf.go:812]: @benarent What do you think?

File: lib/service/cfg.go [hunk 8]
```diff
@@ -445,6 +448,39 @@ type SSHConfig struct {
...
+	// Protocol used to proxy this application. At the moment only HTTPS is
+	// supported.
+	Protocol string `yaml:"protocol"`
+
+	// InternalAddr is the internal address of the application.
+	InternalAddr utils.NetAddr `yaml:"internal_addr"`
+
+	// Public address of the application. This is the address users will access
+	// the application at.
+	PublicAddr utils.NetAddr `yaml:"public_addr"`
```
[the remark this answers is not in the document]
Comment (russjones) [reply lib/service/cfg.go:475]: This is not the address the application will listen on, but the address the application will be reached at.

File: lib/auth/proto/auth.proto [hunk 9]
```diff
@@ -203,6 +203,55 @@ message DeleteUserRequest {
...
 
+// GetAppRequest is the request message to fetch a single application.
+message GetAppRequest {
+    // Namespace is the namespace for the registered application.
+    string Namespace = 1 [ (gogoproto.jsontag) = "namespace" ];
+    // Name is the name of the application being requested.
+    string Name = 2 [ (gogoproto.jsontag) = "name" ];
+    // SkipValidation is used to not use JSON schema validation for
+    // performance reasons.
+    bool SkipValidation = 3 [ (gogoproto.jsontag) = "skip_validation" ];
```
[the remark this answers is not in the document]
Comment (russjones) [reply lib/auth/proto/auth.proto:214]: Will be used in another PR.

File: lib/services/suite/suite.go [hunk 10]
```diff
@@ -356,6 +356,65 @@ func (s *ServicesTestSuite) ServerCRUD(c *check.C) {
...
+	// Create application.
+	app := NewApp("foo", "127.0.0.1:8080", "foo.example.com")
+
+	// Expect not to be returned any applications and trace.NotFound.
+	out, err := s.PresenceS.GetApps(context.Background(), defaults.Namespace)
```
Comment (awly) [inline lib/services/suite/suite.go:382]: nit: create a single context and reuse it throughout the test

File: lib/services/types.proto [hunk 11]
```diff
@@ -10,18 +10,21 @@ option (gogoproto.unmarshaler_all) = true;
...
     int64 LeaseID = 3 [ (gogoproto.jsontag) = "lease_id" ];
-    // Expires is set to update expiry time
+    // Expires is set to update expiry time of the resource.
     google.protobuf.Timestamp Expires = 4 [
         (gogoproto.stdtime) = true,
         (gogoproto.nullable) = false,
         (gogoproto.jsontag) = "expires"
     ];
+    // Type is the type of keep alive, either server or app.
+    string Type = 5 [ (gogoproto.jsontag) = "type" ];
```
Comment (awly) [inline lib/services/types.proto:27]: can this be an `enum`?

File: lib/services/types.proto [hunk 12]
```diff
@@ -156,6 +159,11 @@ message ServerSpecV2 {
     bool UseTunnel = 6 [ (gogoproto.jsontag) = "use_tunnel,omitempty" ];
     // TeleportVersion is the teleport version that the server is running on
     string Version = 7 [ (gogoproto.jsontag) = "version" ];
+    // Protocol used to proxy this application. At the moment only HTTPS is
+    // supported.
+    string Protocol = 8 [ (gogoproto.jsontag) = "protocol,omitempty" ];
```
Comment (awly) [inline lib/services/types.proto:164]: ca this be an `enum`?

File: lib/services/types.proto [hunk 13]
```diff
@@ -156,6 +159,11 @@ message ServerSpecV2 {
...
+    // InternalAddr is the internal address of the application.
+    string InternalAddr = 9 [ (gogoproto.jsontag) = "internal_addr" ];
```
Comment (awly) [inline lib/services/types.proto:166]: `omitempty` here too

File: lib/auth/proto/auth.proto [hunk 14]
```diff
@@ -203,6 +203,55 @@ message DeleteUserRequest {
...
+}
+
+// GetAppsResponse is the response message for all registered applications.
+message GetAppsResponse {
+    // Apps is a slice of services.Server that represent applications.
+    repeated services.ServerV2 Apps = 1 [ (gogoproto.jsontag) = "apps" ];
+}
+
+// UpsertAppRequest is the request message to registered and application.
+message UpsertAppRequest { services.ServerV2 App = 1 [ (gogoproto.jsontag) = "app" ]; }
```
[the remark this answers is not in the document]
Comment (russjones) [reply lib/auth/proto/auth.proto:239]: Unfortunately it's the code formatter doing this in the `make grpc` target.

File: lib/services/server.go [hunk 15]
```diff
@@ -249,6 +260,41 @@ func (s *ServerV2) GetCmdLabels() map[string]CommandLabel {
...
+
+// SetProtocol sets the protocol supported by this server.
+func (s *ServerV2) SetProtocol(protocol string) error {
+	switch protocol {
+	case teleport.ServerProtocolSSH:
+		s.Spec.Protocol = ServerSpecV2_SSH
+	case teleport.ServerProtocolHTTPS:
+		s.Spec.Protocol = ServerSpecV2_HTTPS
+	default:
+		return trace.BadParameter("unknown protocol: %v", protocol)
```
Comment (awly) [inline lib/services/server.go:283]:
```suggestion
		return trace.BadParameter("unknown protocol: %q", protocol)
```

**Review by awly — APPROVED** [review]

**Pull request closed without merging**